### PR TITLE
Fix `defaultValues` getting overwritten

### PR DESCRIPTION
**Pull request by rplopes**

### CODE REVIEW

**Comment by ruiquelhas** [comment]
Is `_.clone` enough or should we use `_.cloneDeep` instead? What kind of information does the function return? Also, we should add one or more tests for this so we can get it merged easily in the main repo.

**Comment by rplopes** [comment]
I took a look at the test suite (in hindsight, I don't know how I didn't do that sooner). I'm stuck on `test/integration/model/create.test.js:781`:

```
  1) [SQLITE] Model create is possible to use functions as default values:
     SequelizeValidationError: string violation: year cannot be an array or an object
      at lib/instance-validator.js:73:14
  From previous event:
      at Promise.then (lib/promise.js:21:17)
      at Context.<anonymous> (test/integration/model/create.test.js:811:53)
```

This test is passing an SQLite function as a default value of the instance, and that isn't working very well with both `_.clone` and `_.cloneDeep`. That is because at that point in time, `valueFn` is:

```
{ fn: '', args: [ { fn: 'date', args: [Object] } ] }
```

(i.e. the SQLite function, still in JSON object format). So by cloning that object, we seem to somehow lose something important, because the value of `year` (the attribute with that default value) will be `undefined`, thus causing that validation error.

Any ideas on why cloning this SQLite function JSON object instead of using it directly makes `year` become `undefined` instead of a string? Is it something important that would make the change in this PR not suitable to be merged?

**Comment by rplopes** [comment]
I've updated these changes to make the tests pass, and added a new unit test with what I consider a sensible behaviour, that would fail under the old code.

In response to the question in my previous comment, that integration test was using a sequelize method as a default value. So `valueFn()._isSequelizeMethod` would return `true`. However, `_isSequelizeMethod` doesn't get cloned on `_.clone()` or `_.cloneDeep()`, so `_.cloneDeep(valueFn())._isInstanceMethod` is `undefined`. Later, in validations, because `year` was meant to be a string, its value `{ fn: '', args: [ { fn: 'date', args: [Object] } ] }` was being parsed as an object (this explains the error message: `SequelizeValidationError: string violation: year cannot be an array or an object`). This validation doesn't apply to sequelize methods, hence the difference in behaviour.

In summary, the issue is fixed and I'm again submitting this to be merged.

**Comment by ruiquelhas** [comment]
Have you tried using `_.mapValues()` without cloning?

**Comment by rplopes** [comment]
@ruiquelhas Yes, and it behaves just like `_.each()` was behaving. Which is ok, I guess, because apparently we don't want to clone values when they're sequelize methods (their `_isSequelizeMethod` attribute isn't cloned, which results in errors later).

File: test/unit/instance/build.test.js [hunk 1]
```diff
@@ -71,5 +71,21 @@ describe(Support.getTestDialectTeaser('Instance'), function() {
...
       expect(instance.get('id')).to.be.ok;
     });
+
+    it('should change values for individual instances only', function () {
+      var Model = current.define('Model', {
+        model_attribute: {
+          type: DataTypes.JSONB,
+          defaultValue: { key: 'default value' }
+        }
+      })
```
Comment (nunofgs) [inline test/unit/instance/build.test.js:81]: Missing semicolon.

File: test/unit/instance/build.test.js [hunk 2]
```diff
@@ -71,5 +71,21 @@ describe(Support.getTestDialectTeaser('Instance'), function() {
       expect(instance.get('id')).not.to.be.undefined;
       expect(instance.get('id')).to.be.ok;
     });
+
+    it('should change values for individual instances only', function () {
+      var Model = current.define('Model', {
+        model_attribute: {
```
Comment (nunofgs) [inline test/unit/instance/build.test.js]: I suggest using `data` as the field name.

File: test/unit/instance/build.test.js [hunk 3]
```diff
@@ -71,5 +71,21 @@ describe(Support.getTestDialectTeaser('Instance'), function() {
...
+          type: DataTypes.JSONB,
+          defaultValue: { key: 'default value' }
```
Comment (nunofgs) [inline test/unit/instance/build.test.js]: Suggestion for defaultValue: `{ foo: 'bar' }`

File: test/unit/instance/build.test.js [hunk 4]
```diff
@@ -71,5 +71,21 @@ describe(Support.getTestDialectTeaser('Instance'), function() {
...
+        }
+      })
+        , instance;
```
Comment (nunofgs) [inline test/unit/instance/build.test.js:82]: Use `var`.

File: test/unit/instance/build.test.js [hunk 5]
```diff
@@ -71,5 +71,21 @@ describe(Support.getTestDialectTeaser('Instance'), function() {
...
+
+      instance = Model.build();
```
Comment (nunofgs) [inline test/unit/instance/build.test.js:84]: This can be moved to the assignment above.

File: test/unit/instance/build.test.js [hunk 6]
```diff
@@ -71,5 +71,21 @@ describe(Support.getTestDialectTeaser('Instance'), function() {
...
+      instance.model_attribute.key = 'new value';
```
Comment (nunofgs) [inline test/unit/instance/build.test.js]: Suggestion for value: `biz`

File: test/unit/instance/build.test.js [hunk 7]
```diff
@@ -71,5 +71,21 @@ describe(Support.getTestDialectTeaser('Instance'), function() {
       expect(instance.get('id')).not.to.be.undefined;
       expect(instance.get('id')).to.be.ok;
     });
+
+    it('should change values for individual instances only', function () {
```
Comment (nunofgs) [inline test/unit/instance/build.test.js]: Suggestion for test description: `it('should clone the default values')` (or something along those lines).

File: lib/instance.js [hunk 8]
```diff
@@ -22,10 +22,9 @@ var initValues = function(values, options) {
     defaults = {};
 
     if (this.Model._hasDefaultValues) {
-      Utils._.each(this.Model._defaultValues, function(valueFn, key) {
-        if (defaults[key] === undefined) {
-          defaults[key] = valueFn();
-        }
+      defaults = _.mapValues(this.Model._defaultValues, function(valueFn) {
+        var value = valueFn();
+        return value._isSequelizeMethod ? value : _.cloneDeep(value);
```
Comment (nunofgs) [inline lib/instance.js:27]: The `_isSequelizeMethod` check does not work for me. Even if it did, I still think we should "cloneDeep" regardless if a function was given or not. This also protects users from passing a function and returning the same object reference.

Suggested implementation:

``` js
      defaults = _.mapValues(this.Model._defaultValues, function(fn) {
        return _.cloneDeep(fn());
      });
```

Comment (rplopes) [inline lib/instance.js:27]: I tried `_.clone` and `_.cloneDeep` before, but that would cause an issue with sequelize methods, making an integration test fail and potentially introducing regressions. More info in https://github.com/seegno-forks/sequelize/pull/3#issuecomment-166697719 and https://github.com/seegno-forks/sequelize/pull/3#issuecomment-167829881

File: test/unit/instance/build.test.js (hunk 1 above)
Comment (rplopes) [inline test/unit/instance/build.test.js:81]: Here, I'm following the convention of this file (see e.g. test above). I believe it's better to follow that convention, specially if we plan to submit this upstream later.

File: test/unit/instance/build.test.js (hunk 4 above)
Comment (rplopes) [inline test/unit/instance/build.test.js:82]: Same as above.

File: test/unit/instance/build.test.js (hunk 5 above)
Comment (rplopes) [inline test/unit/instance/build.test.js:84]: Here, I'm following the convention of this file (see e.g. test above). I believe it's better to follow that convention, specially if we plan to submit this upstream later.

File: lib/instance.js (hunk 8 above)
Comment (nunofgs) [inline lib/instance.js:27]: Can you investigate further please? I was able to run all unit and integration tests.

Comment (rplopes) [inline lib/instance.js:27]: Using that code, I get an integration test for SQLite failing, when running `npm t`:

```
  1) [SQLITE] Model create is possible to use functions as default values:
     SequelizeValidationError: string violation: year cannot be an array or an object
      at lib/instance-validator.js:72:14
  From previous event:
      at Promise.then (lib/promise.js:21:17)
      at Context.<anonymous> (test/integration/model/create.test.js:811:53)
```

It's the same error I was getting before. From my previous research, it seems to be because `valueFn()._isSequelizeMethod` isn't being correctly cloned. If this is passing on your machine, it might be a bug with some software version on mine.

Comment (nunofgs) [inline lib/instance.js:27]: Nevermind, it was a previous edit of mine. LGTM.
